### PR TITLE
Re-implement clspv-opt as a standalone driver

### LINKED ISSUE #307: clsvp-opt should support windows
The current implementation of clsvp-opt (#283) does not support Windows because it is implemented as a plugin to LLVM's `opt`.  This is difficult to do out of tree because of the way LLVM and Windows DLL libraries work (http://go/so/41943353).

Extend clsvp-opt so it knows how to read/write LLVM IR without having to use `opt`.

### PR DESCRIPTION
**Pull request by dnovillo**
This fixes #307 and #322.

It re-implements clspv-opt as a standalone driver.  The initial
implementation used the plugin facility for LLVM opt, but this is
difficult to implement in Windows (#307).

The driver only supports LLVM scalar transformations and all of clspv
transformations.  To do this, passes must be declared using LLVM's
INITIALIZE_PASS macro.  Additionally, the function
initializeClspvPasses() must include a call to the initialize*()
function generated by INITIALIZE_PASS.

This change also adds tests for all the existing passes in clspv.  It
does not check for specific transformations.  It checks that the flags
are accepted by clspv-opt.

Finally, this change reduces the build time for LLVM by building no
target backends by default.

### CODE REVIEW

File: include/clspv/Passes.h [hunk 1]
```diff
@@ -25,6 +25,47 @@ class ModulePass;
...
 template <typename T> class ArrayRef;
+
+// Clspv pass initializers.
+class PassRegistry;
+
+// Initialize all clspv passes.
+void initializeClspvPasses(PassRegistry &);
+
+// Individual pass initializers.
+void initializeAllocateDescriptorsPassPass(PassRegistry &);
```
Comment (alan-baker) [inline include/clspv/Passes.h]: Do these individual pass initializers need to be exposed in the public header?

File: lib/AllocateDescriptorsPass.cpp [hunk 2]
```diff
@@ -160,8 +160,8 @@ class AllocateDescriptorsPass final : public ModulePass {
 } // namespace
 
 char AllocateDescriptorsPass::ID = 0;
-static RegisterPass<AllocateDescriptorsPass> X("AllocateDescriptorsPass",
-                                               "Allocate resource descriptors");
+INITIALIZE_PASS(AllocateDescriptorsPass, "AllocateDescriptorsPass",
+                "Allocate resource descriptors", false, false)
```
Comment (alan-baker) [inline lib/AllocateDescriptorsPass.cpp]: One of the falses is whether this is an analysis. What is the other?

File: lib/Option.cpp [hunk 3]
```diff
@@ -152,6 +155,46 @@ static llvm::cl::opt<bool>
               llvm::cl::desc("Enable experimental C++ support"));
 } // namespace
 
+namespace llvm {
+
+void initializeClspvPasses(PassRegistry &r) {
```
Comment (alan-baker) [inline lib/Option.cpp]: This is an odd file to contain the initialization. I'd suggest making a Passes.cpp.

File: include/clspv/Passes.h (hunk 1 above)
Comment (dnovillo) [reply include/clspv/Passes.h]: In some header available to `clspv-opt`.

File: lib/AllocateDescriptorsPass.cpp (hunk 2 above)
Comment (dnovillo) [reply lib/AllocateDescriptorsPass.cpp]: Whether it's a CFG-only pass.

File: lib/Option.cpp (hunk 3 above)
Comment (dnovillo) [reply lib/Option.cpp]: Agreed.  I was going to suggest that and then forgot.

File: include/clspv/Passes.h (hunk 1 above)
Comment (dnovillo) [reply include/clspv/Passes.h]: Oh, and available to the passes too.  The macro expands to code that defines the `initialize*` function, so the declaration needs to be available.  `clspv/Passes.h` seemed the best home for these.

Comment (alan-baker) [reply include/clspv/Passes.h]: Doesn't just the overall initializer need to be available to clspv-opt? Could we have a header local to lib/ that declares the individual pass initializers?

Comment (dnovillo) [reply include/clspv/Passes.h]: That should work.  Yes.

**Comment by dnovillo** [comment]
Added `lib/Passes.{cpp,h}`.  PTAL.

File: lib/AllocateDescriptorsPass.cpp [hunk 4]
```diff
@@ -26,9 +26,9 @@
 #include "llvm/Support/CommandLine.h"
 #include "llvm/Support/raw_ostream.h"
 
+#include "Passes.h"
```
Comment (alan-baker) [inline lib/AllocateDescriptorsPass.cpp]: This (and other instances) no longer match the style of other local directory includes like ArgKind.h and Constants.h

Comment (dnovillo) [reply lib/AllocateDescriptorsPass.cpp]: What do you mean?

Comment (dnovillo) [reply lib/AllocateDescriptorsPass.cpp]: OK, fixed.  PTAL.

File: CMakeLists.txt [hunk 5]
```diff
@@ -69,15 +69,7 @@ set(CMAKE_POSITION_INDEPENDENT_CODE ON)
...
-if (NOT MSVC)
-  set(ENABLE_CLSPV_OPT ON)
-else()
-  # clspv-opt is not supported on Windows, because LLVM opt does not support
-  # out-of-tree plugins on Windows (http://go/so/41943353).
-  # TODO(dnovillo): Need a different implementation that does not rely on
-  #                 LLVM opt.
-  set(ENABLE_CLSPV_OPT OFF)
-endif()
+option(ENABLE_CLSPV_OPT "Enable the clspv-opt driver." ON)
```
Comment (jgavris) [inline CMakeLists.txt]: I was trying an experiment to generate some code without optimizations, does this flag affect the `clspv` target at the moment? I built with this on and off, I get identical output for `clspv`.

Comment (alan-baker) [reply CMakeLists.txt]: This controls whether to build the standalone executable clspv-opt. It now builds on all platforms by default. The executable just lets us run individual passes on IR (for testing purposes). You can -O0 as an input to clspv to disable optimizations, but there are no guarantees it generates correct code.

**Pull request merged**